### PR TITLE
Add ngramClassify and ngramClassifyUTF8 functions

### PR DESCRIPTION
**Pull request by 0442A403**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ```ngramClassify``` and ```ngramClassifyUTF8``` functions

### Documentation entry for user-facing changes

- [x] Documentation is written
- [x] Tests are written

Task is described at #42194 as "Text classification with ngram models".

This PR adds new function ```ngramClassify``` and ```ngramClassifyUTF8``` which implements naive Bayes classifier. These functions are a easy way to classify text very quickly (sacrificing quality of course). 

For example you need to find offensive comments about football and hockey then you can query:
```sql
SELECT comment_text
FROM comments
WHERE ngramClassify('offensive', comment_text) = 'positive'
AND ngramClassify('sport', comment_text) in ('football', 'hockey');
```
So it is based on models. In example above there are 2 models: ```offensive``` and ```sport```. These models are based on texts that must represent all its class. 
Models are configurated in config. So for query above there could be the following config:
```xml
<ngram_classifiers>
    <ngram_classifier>
        <ngram_classifier_name>offensive</ngram_classifier_name>
        <ngram_classes>
            <ngram_class>
                <ngram_class_name>positive</ngram_class_name>
                <ngram_class_path>/etc/offensive.txt</ngram_class_path>
            </ngram_class>
            <ngram_class>
                <ngram_class_name>negative</ngram_class_name>
                <ngram_class_path>/etc/non-offensive.txt</ngram_class_path>
            </ngram_class>
        </ngram_classes>
    </ngram_classifier>
    <ngram_classifier>
        <ngram_classifier_name>sport</ngram_classifier_name>
        <ngram_classes>
            <ngram_class>
                <ngram_class_name>football</ngram_class_name>
                <ngram_class_path>/etc/football.txt</ngram_class_path>
            </ngram_class>
            <ngram_class>
                <ngram_class_name>basketball</ngram_class_name>
                <ngram_class_path>/etc/backetball.txt</ngram_class_path>
            </ngram_class>
            <ngram_class>
                <ngram_class_name>hockey</ngram_class_name>
                <ngram_class_path>/etc/hockey.txt</ngram_class_path>
            </ngram_class>
            <ngram_class>
                <ngram_class_name>chess</ngram_class_name>
                <ngram_class_path>/etc/chess.txt</ngram_class_path>
            </ngram_class>
        </ngram_classes>
    </ngram_classifier>
</ngram_classifiers>
```
Template for configuration is located on ```programs/server/config.d/ngram_classifiers.xml```.

In the sequal of this PR functionality can expand up to next features:
- Loading models from the CSV file without processing it after reading
- Dumping processed models for the purpose of previous point
- Downloading models with FTP
- Publications some frequently asked models on ClickHouse resources so everyone could just type ```ngramClassify('clickhouse.offensive', text)```
- Other Bayes classifiers (could be found [here](https://scikit-learn.org/stable/modules/naive_bayes.html#multinomial-naive-bayes))

### CODE REVIEW

File: src/Functions/FunctionsStringSimilarity.h [hunk 1]
```diff
@@ -132,4 +133,163 @@ class FunctionsStringSimilarity : public IFunction
...
+        }
+
+        const String & model_name = col_needle_const->getValue<String>();
+        const auto& model_it = string_similarity_models.find(model_name);
+        if (model_it == string_similarity_models.end()) {
+            throw Exception("There is no " + model_name + " string similarity model", ErrorCodes::BAD_ARGUMENTS);
+        }
+        const auto& model = model_it->second;
+
+        const String positive_needle = readModel(model.path_to_positive);
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsStringSimilarity.h:216]: The models are currently loaded again for each block.
Instead, they should be managed by Context, and reloaded similarly to Dictionaries.

File: src/Functions/FunctionsStringSimilarity.h [hunk 2]
```diff
@@ -132,4 +133,163 @@ class FunctionsStringSimilarity : public IFunction
...
+
+        String model;
+        model.reserve((in.available()));
+        readStringUntilEOF(model, in);
+
+        in.close();
+        return model;
+    }
+};
+
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsStringSimilarity.h:294]: Missing tests.

File: src/Functions/FunctionsStringSimilarity.h [hunk 3]
```diff
@@ -132,4 +133,163 @@ class FunctionsStringSimilarity : public IFunction
     }
 };
 
+struct StringSimilarityModel {
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsStringSimilarity.h:136]: Let's use a dedicated file for the new code.

**Comment by alexey-milovidov** [comment]
It does not build.

**Comment by alexey-milovidov** [comment]
Missing the analysis of the algorithm (for charset or language classification).

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 1105d6174b5da36fb475ff78b4e1936df9ae02f6 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48984/1105d6174b5da36fb475ff78b4e1936df9ae02f6/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: tests/config/offensive-ngrams.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+offensive angry war violance evil adversity bad break bummer calamity cataclysm catastrophe disaster downer drag evil hard knocks hard times hardship jam misery misfortune mishap storm cloud tragedy tribulation trouble unpleasantnes
```
Comment (alexey-milovidov) [inline tests/config/offensive-ngrams.txt:1]: This is not as discussed. The model file should contain (class_id, ngram, count) triples.

**Pull request closed without merging**